### ISSUE TITLE
ElasticSearch Allows Decimal For Type BYTE

### ISSUE THREAD

**Issue by threejeez** [elastic/elasticsearch#16514]
ElasticSearch documentation states:

`The number types have the same ranges as corresponding Java types`

[Java types](http://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html) is linked and states the following:

`The byte data type is an 8-bit signed two's complement integer.`

Yet when I have a field mapped as type "byte" I can store a decimal when indexing it as a number. When I index the number as a string it gives me the correct error.

**Mapping**

```
POST /my_index
{
    "mappings": {
        "my_type": {
            "properties": {
                "my_byte": {
                    "type": "byte"
                }
            }
        }
    }
}
```

**Index as number**

```
POST /my_index/my_type
{
   "my_byte":  98.6
}
```
```
{
   "_index": "my_index",
   "_type": "my_type",
   "_id": "AVLCEWrzgBH0py33OQn-",
   "_version": 1,
   "_shards": {
      "total": 2,
      "successful": 1,
      "failed": 0
   },
   "created": true
}
```

**GET request and response**

```
GET /my_index/my_type/AVLCEWrzgBH0py33OQn-
```
```
{
   "_index": "my_index",
   "_type": "my_type",
   "_id": "AVLCEWrzgBH0py33OQn-",
   "_version": 1,
   "found": true,
   "_source": {
      "my_byte": 98.6
   }
}
```

**Unsuccessful INDEX as string (correct behavior)**

```
POST /my_index/my_type
{
    "my_byte":"98.6"
}
```
```
{
   "error": {
      "root_cause": [
         {
            "type": "mapper_parsing_exception",
            "reason": "failed to parse [my_byte]"
         }
      ],
      "type": "mapper_parsing_exception",
      "reason": "failed to parse [my_byte]",
      "caused_by": {
         "type": "number_format_exception",
         "reason": "For input string: \"98.6\""
      }
   },
   "status": 400
}
```

**Comment by rjernst** [comment]
The behavior you see is actually expected, at least when passing in the decimal. The `coerce` setting of numeric fields defaults to true, and part of the definition of coerce is `Floating points will be truncated for integer values.` . See https://www.elastic.co/guide/en/elasticsearch/reference/2.0/coerce.html.

I believe the bug here is actually in what you are seeing for string behavior. The current code does a "coerce check" in the string case, but that only guards before it tries to call Integer.parseInt. However, in the coerce case, Double.parseDouble should be called, and then longValue(). @clintongormley thoughts?

**Comment by threejeez** [comment]
@rjernst Thanks for your response. Based on the link you provided - and considering the mapping above - coerce should be set to the default value of true. But it doesn't seem to be coercing anything, evidenced by the result I'm getting when retrieving the document:

```
   "_source": {
      "my_byte": 98.6 
   }
```

The value I'm getting back **is a double**. If it's being coerced at all (while indexed or retrieved) I would expect to get back the truncated value of `98`.

**Comment by rjernst** [comment]
The value is coerced before indexing, but source is never changed (imagine you want to reindex as a float in the future).  If you want source to be truncated, you will need to do that yourself when ingesting the docs.

**Comment by threejeez** [comment]
Your point about _source makes perfect sense. Thank you for clarifying. Can you please point out how to get the coerced (int) value? I assumed it would be by specifying the `fields` parameter, but I'm seeing a difference between fields returned when doing a GET vs a SEARCH:

This: ```curl search.keenhome.io:9200/my_index/my_type/_search?fields=my_byte```
Yeilds a double: ```"fields":{"my_byte":[98.6]}```

Whereas this: ```curl search.keenhome.io:9200/my_index/my_type/AVLCEWrzgBH0py33OQn-?fields=my_byte```
Yields an int: ```"fields":{"my_byte":[98]}```

Is this expected behavior?

**Comment by clintongormley** [comment]
@rjernst makes sense to me.  Perhaps this should be folded into a bigger cleanup as in https://github.com/elastic/elasticsearch/issues/11513

**Comment by jpountz** [comment]
This is fixed in 5.0.
